File: src/main/java/pers/neige/neigeitems/action/container/ActionTrigger.java
package pers.neige.neigeitems.action.container;

import lombok.NonNull;
import org.bukkit.configuration.ConfigurationSection;
import pers.neige.neigeitems.action.Action;
import pers.neige.neigeitems.action.ActionContext;
import pers.neige.neigeitems.action.ActionResult;
import pers.neige.neigeitems.manager.BaseActionManager;
import pers.neige.neigeitems.utils.SchedulerUtils;

import java.util.concurrent.CompletableFuture;

public class ActionTrigger {
    private final @NonNull BaseActionManager actionManager;
    private final @NonNull String type;
    private @NonNull Action actions;
    private @NonNull Action async;
    private @NonNull Action sync;

    public ActionTrigger(
        @NonNull BaseActionManager actionManager,
        @NonNull String type,
        @NonNull ConfigurationSection config
    ) {
        this.actionManager = actionManager;
        this.type = type;
        this.actions = actionManager.compile(config.get("actions"));
        this.async = actionManager.compile(config.get("async"));
        this.sync = actionManager.compile(config.get("sync"));
    }

    public ActionTrigger(
        @NonNull BaseActionManager actionManager,
        @NonNull String type,
        @NonNull Action actions
    ) {
        this.actionManager = actionManager;
        this.type = type;
        this.actions = actions;
        this.async = actionManager.NULL_ACTION;
        this.sync = actionManager.NULL_ACTION;
    }

    public @NonNull BaseActionManager getActionManager() {
        return actionManager;
    }

    public @NonNull String getType() {
        return type;
    }

    public @NonNull Action getActions() {
        return actions;
    }

    public void setActions(@NonNull Action actions) {
        this.actions = actions;
    }

    public @NonNull Action getAsync() {
        return async;
    }

    public void setAsync(@NonNull Action async) {
        this.async = async;
    }

    public @NonNull Action getSync() {
        return sync;
    }

    public void setSync(@NonNull Action sync) {
        this.sync = sync;
    }

    public @NonNull CompletableFuture<ActionResult> run(
        @NonNull ActionContext context
    ) {
        async(context);
        sync(context);
        return actions.eval(context);
    }

    public void async(
        @NonNull ActionContext context
    ) {
        SchedulerUtils.async(actionManager.getPlugin(), () -> {
            async.eval(context.clone());
        });
    }

    public void sync(
        @NonNull ActionContext context
    ) {
        SchedulerUtils.sync(actionManager.getPlugin(), () -> {
            sync.eval(context.clone());
        });
    }
}

File: src/main/kotlin/pers/neige/neigeitems/item/action/ActionTrigger.kt
package pers.neige.neigeitems.item.action

import org.bukkit.configuration.ConfigurationSection
import pers.neige.neigeitems.action.Action
import pers.neige.neigeitems.action.ActionContext
import pers.neige.neigeitems.manager.ActionManager
import pers.neige.neigeitems.utils.SchedulerUtils

/**
 * 物品动作触发器
 *
 * @property id 物品ID
 * @property type 触发器类型
 * @property config 触发器配置
 */
class ActionTrigger(val id: String, val type: String, val config: ConfigurationSection) {
    /**
     * 获取物品使用冷却
     */
    val cooldown = config.getString("cooldown", "1000")

    /**
     * 获取tick型触发器触发间隔
     */
    val tick = config.getString("tick", "10")

    /**
     * 获取物品冷却组ID
     */
    val group: String = config.getString("group") ?: "${type}_$id"

    /**
     * 获取物品消耗信息
     */
    val consume: ConsumeInfo? = config.getConfigurationSection("consume")?.let { ConsumeInfo(it) }

    /**
     * 获取异步动作信息
     */
    val actions: Action = ActionManager.compile(config.get("actions"))

    /**
     * 获取异步动作信息
     */
    val async: Action = ActionManager.compile(config.get("async"))

    /**
     * 获取同步动作信息
     */
    val sync: Action = ActionManager.compile(config.get("sync"))

    /**
     * 运行异步物品动作
     *
     * @param context 动作上下文
     */
    fun run(
        context: ActionContext
    ) {
        async(context)
        sync(context)
    }

    /**
     * 运行异步物品动作
     *
     * @param context 动作上下文
     */
    fun async(
        context: ActionContext
    ) {
        SchedulerUtils.async {
            actions.eval(context)
            async.eval(context)
        }
    }

    /**
     * 运行同步物品动作
     *
     * @param context 动作上下文
     */
    fun sync(
        context: ActionContext
    ) {
        SchedulerUtils.sync {
            sync.eval(context)
        }
    }
}